library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java

```diff
@@ -286,271 +146,233 @@ public AbstractJdbcPollInputOperator()
-      LOG.info("Statement when re-initialized {}", ps.toString());
-      dbPoller = new Thread(new DBPoller());
-      dbPoller.start();
-    }
   }
 
   @Override
   public void emitTuples()
   {
-    if (isReplayed) {
+    if (currentWindowId <= windowManager.getLargestRecoveryWindow()) {
       return;
     }
-
-    List<T> tuples;
-
-    if ((tuples = emitQueue.poll()) != null) {
-      for (Object tuple : tuples) {
-        if (lower == null) {
-          lower = tuple.toString();
-        }
-        upper = tuple.toString();
-        outputPort.emit((T)tuple);
+    int pollSize = (emitQueue.size() < batchSize) ? emitQueue.size() : batchSize;
+    while (pollSize-- > 0) {
+      T obj = emitQueue.poll();
+      if (obj != null) {
+        emitTuple(obj);
       }
+      lastEmittedRecord++;
     }
   }
 
+  protected void emitTuple(T obj)
+  {
+    outputPort.emit(obj);
+  }
+
   @Override
   public void endWindow()
   {
     try {
       if (currentWindowId > windowManager.getLargestRecoveryWindow()) {
-        if (currentWindowRecoveryState == null) {
-          currentWindowRecoveryState = new MutablePair<String, String>();
-        }
-        if (lower != null && upper != null) {
-          previousUpperBound = upper;
-          isPolled = true;
-        }
-        MutablePair<String, String> windowBoundaryPair = new MutablePair<>(lower, upper);
-        currentWindowRecoveryState = windowBoundaryPair;
+        currentWindowRecoveryState = new MutablePair<>(lowerBound, lastEmittedRecord);
         windowManager.save(currentWindowRecoveryState, operatorId, currentWindowId);
       }
     } catch (IOException e) {
       throw new RuntimeException("saving recovery", e);
     }
-    currentWindowRecoveryState = new MutablePair<>();
-  }
-
-  public int getPartitionCount()
-  {
-    return partitionCount;
+    if (threadException != null) {
+      store.disconnect();
+      DTThrowable.rethrow(threadException.get());
+    }
   }
 
-  public void setPartitionCount(int partitionCount)
+  @Override
+  public void deactivate()
   {
-    this.partitionCount = partitionCount;
+    scanService.shutdownNow();
+    store.disconnect();
   }
 
-  @Override
-  public void activate(Context cntxt)
+  protected void pollRecords()
   {
-    if (context.getValue(OperatorContext.ACTIVATION_WINDOW_ID) != Stateless.WINDOW_ID
-        && context.getValue(OperatorContext.ACTIVATION_WINDOW_ID) < windowManager.getLargestRecoveryWindow()) {
-      // If it is a replay state, don't start any threads here
+    if (isPolled) {
       return;
     }
-  }
-
-  @Override
-  public void deactivate()
-  {
     try {
-      if (dbPoller != null && dbPoller.isAlive()) {
-        dbPoller.interrupt();
-        dbPoller.join();
+      ps.setFetchSize(getFetchSize());
+      ResultSet result = ps.executeQuery();
+      if (result.next()) {
+        do {
+          emitQueue.offer(getTuple(result));
```

Comment:
Here, we need to check if ```offer()``` returns ```false```. If so, keep trying (after sleeping for a while) until it returns success.